docs/src/lucet-runtime/killswitch.md

```diff
@@ -0,0 +1,346 @@
+}
+
+// wait for the KillSwitch-firing thread to complete
+t.join().unwrap();
+```
+
+## Implementation
+
+As this section discusses implementation details of `lucet-runtime`, it
+will refer to structures and enums that are unexported.  Since there are
+no `lucet-runtime` paths for these items, this section will occasionally
+use fully-qualified paths to the `lucet_runtime_internals` path where.
+they are originally defined.
+
+As much as is possible, `KillSwitch` tries to be self-contained; no members are
+public, and it tries to avoid leaking details of its state-keeping into public
+interfaces. Currently, `lucet-runtime` is heavily dependent on POSIX
+thread-directed signals to implement guest timeouts - for non-POSIX platforms
+alternate implementations may be plausible, but need careful consideration of
+the race conditions that can arise from other platform-specific functionality.
+
+`KillSwitch` fundamentally relies on two piece of state for safe operation,
+which are encapsulated in a `KillState` held by the `lucet_runtime::Instance`
+it terminates:
+* `execution_domain`, a `Domain` that describes the kind of execution that is
+  currently happening in the `Instance`. This is kept in a `Mutex` since in
+  many cases it will need to be accessed either by a `KillSwitch` or
+  `KillState`, and for the duration either are considering the domain, it must
+  block other users.
+* `terminable`, an `AtomicBool` that indicates if the `Instance` may stop
+  executing.
+
+The astute observer may notice that `Lucet` contains both an instance `Domain`
+and an instance `State`. These discuss different aspects of the instance's
+lifecycle: `Domain` describes what the instance is doing right now, where
+`State` describes the state of the instance as visible outside
+`lucet_runtime`. This is fundamentally the reason why `State` does not and
+cannot describe a "making a hostcall" variant - this is not something
+`lucet_runtime` wants to expose, and at the moment we don't intend to make any
+commitments about being able to query instance-internal state like this.
+Additionally, this is why `Domain` only expresses that an instance has
+_stopped_ running, not if it successfully or unsuccessfully exited. TODO:
+there's even an argument that `Terminated` vs `Cancelled` is a stronger
+distinction than `Domain` needs to make!
+
+### Termination Mechanisms
+
+At a high level,`KillSwitch` picks one of several mechanisms to terminate an
+instance.  The choice of which mechanism depends on what kind of code is
+running in the guest, indicated by the `instance::execution::Domain` enum.
+
+`Domain::Pending` is the easiest domain to stop execution in: we simply update
+the execution domain to `Cancelled` so when the instance is run we know that we
+in fact should not run it.
+
+`Domain::Guest` is likely to be the next most-immediate termination form, where
+`lucet_runtime` will send a thread-directed `SIGARLM` to the thread running the
```

Comment:
```suggestion
`Domain::Guest` is likely to be the next most-immediate termination form.
`lucet_runtime` will send a thread-directed `SIGARLM` to the thread running the
```
Rather than describing this in terms of immediacy, I'd list this case before the cancellation case. My reasoning here is that I think this is what I would think of as our "base" case. Other situations are edge cases due to the many devils in the details, but _this_ case is the one I would immediately point someone to if they were attempting to learn about our signalling.

Reply:
The grammar suggestion above is a small nit, in the interest of short digestible sentences.

Reply:
I like your point about which mechanism people will likely _want_ to see. I'll reorder these items.

(also thank you for calling out my run-ons. there are many.)